CLI

aaronfay opened the pull request:
Opening this up for early review -- I've been tinkering with `yargs` a bit to see how it would fit as a wrapper for an initial CLI implementation, and I think so far it would work pretty good! 

A couple cool points worth noting:

 * CLI commands are really just a thin wrapper around existing Node.js client methods
 * `yargs` supports folder-based commands structure to help keep things organized nicely
 * any options input automatically is compatible with our Node.js client `{options}` so minimal processing is required
 * we're not limited to the options we define, so user variables (like `--my-cool-var=foobar`) will automatically be passed as `{{ myCoolVar }}`
 * `yargs` naturally supports combining multiple options into an array (eg: `--browser chrome --browser firefox` becomes `{browser: ['chrome', 'firefox']}`)

We also get the cool help output at every level:

```
❯ ghost-inspector -h
ghost-inspector <command>

Commands:
  ghost-inspector duplicate-test <testId>          Duplicate a test.
  ghost-inspector execute-test <testId> [options]  Execute a test with the provided
                                         options.

Options:
  --version  Show version number                                       [boolean]
  --apiKey   Your Ghost Inspector API key.                            [required]
  --help     Show help                                                 [boolean]

Not enough non-option arguments: got 0, need at least 1
```

```
❯ ghost-inspector execute-test -h
ghost-inspector execute-test <testId> [options]

Execute a test with the provided options.

Options:
  --version     Show version number                                    [boolean]
  --apiKey      Your Ghost Inspector API key.                         [required]
  --help        Show help                                              [boolean]
  --browser     Alternate browser to use for this execution. The following
                options are available: chrome (Latest version), chrome-
                (Specific version of Chrome, for example chrome-83), firefox
                (Latest version), firefox- (Specific version of Firefox, for
                example firefox-77)
  --start-url   Alternate start URL to use for this execution only
  --user-agent  Alternate user agent to use for this execution only
  --region      Geo-location for test execution. The following options are
                available: us-east-1 (default), us-west-1, ca-central-1,
                eu-central-1, eu-west-1, eu-west-2, eu-west-3, eu-north-1,
                me-south-1, ap-east-1, ap-northeast-1, ap-northeast-2,
                ap-southeast-1, ap-southeast-2, ap-south-1, sa-east-1
  --viewport    Alternate screen size to use for this execution only. This
                should be a string formatted as {width}x{height}, for example
                1024x768. Will be ignored if screenshot comparison or visual
                capture is disabled.
  --immediate                                                          [boolean]

Not enough non-option arguments: got 0, need at least 1
```

I know we wanted to have a little more discussion about the CLI so I didn't dive too far into this, mostly wanted to use it as an excuse to learn `yargs` a bit and see what it can do.

Lend me your thoughts! 👂🏻

aaronfay commented:
To test this out, you can pull down this branch and run 

```
npm install -g .
ghost-inspector -h
```

I've also set it up so the `apiKey` can be passed in as an option or in the environment, so both of these will work:

```
❯ GHOST_INSPECTOR_API_KEY=xxx ghost-inspector execute-test 1234 --browser firefox
❯ ghost-inspector execute-test 1234 --apiKey xxx --browser firefox
```

justinklemm commented:
Wowza, nice! I didn't take a good look yet, but we've got to go with either `ghost-inspector` or `ghostinspector` for the binary name. `ghost` is just too common. It's also a [popular blogging library](https://ghost.org/) that may already have a CLI at this point.

simpixelated reviewed (commented):
This is awesome! One small nitpick is that the casing is inconsistent. e.g. `apiKey`, `testid`, and `start-url`. I would expect `api-key` and `test-id` or `startUrl`.

aaronfay commented:
@simpixelated good call, I hadn't realized I'd mixed them up 😂 . `yargs` will actually let you do both and either, and when you pass in `--apiKey xxx` you'll get `{apiKey: 'xxx', 'api-key': 'xxx'}` regardless. I'll clean those up in the meantime 👍🏻

justinklemm commented:
Yea, this looks really good! I think a good start would be list/get/execute commands for suites/tests/results. That would make this really useful off the bat. We could either roll that out and circle back to add commands for everything else -- or if you think it's easy enough to add everything our API offers right away, we can do that.

aaronfay commented:
This is largely ready for feedback 🙏🏻  (short of testing & docs). 

Some notes:

 * There is a bit of duplication in the way each of the commands is implemented (parse args, run method, catch result, print output). This could probably be abstracted to reduce duplication but that felt a little overkill right now. We could always revisit it later if it becomes problematic to maintain.
 * I've added a number of methods that were missing to the Node.js client:
    * `createFolder`
    * `updateFolder`
    * `getAllRunningTests`
    * `createSuite`
    * `updateSuite`
    * `deleteTest`
    * `upateTest`
 * I've added two new methods for downloading, `downloadSuiteJsonBundled` and `downloadTestJsonBundled`. I was originally thinking to add the `includeImports` option to `downloadSuiteJson` but the existing methods don't currently accept any options, and felt messier to fuss with the signatures to check whether we're passing options or a callback. I could be convinced that that is in fact the better approach.

justinklemm commented on bin/commands/suite/import-test.js:
```diff
@@ -0,0 +1,19 @@
+const helpers = require('../../helpers')
+
+// TODO: node.js client says imports still support Selenium IDE v1 format, should we remove this?
```
I believe the [API supports both .html and .json](https://ghostinspector.com/docs/api/suites/#import-test). We should continue to allow both.

Maybe you're thinking of the Selenium v2 Builder (.json) format? I don't think we ever supported import of those files; just export.

justinklemm reviewed (commented):
1. This looks good and I'm fine with the duplicate.
2. Thanks for adding these!
3.  I think we should keep the existing function and add support for `options`. It shouldn't be that hard to check for options or a callback. Creating a whole separate function for what is pretty clearly an option doesn't feel correct. It'll also put us in a very weird spot if we add more options to the download feature. Imagine if we add an another option like `zipOutput` or something... Will we duplicate the functions into another set of new functions again? We should just avoid it and stuck to the existing.pattern.

aaronfay commented:
From the CLI standpoint, this is ready to go. There is one remaining issue that came up while testing which is actually an issue with the Node.js client  when downloading a suite with multiple tests. I'm pretty sure this issue has always been there, so I'm going to write it up as a separate issue and come back to it.

justinklemm commented on README.md:
````diff
@@ -2,13 +2,84 @@
+
+❯ ghost-inspector test execute --help
+...
+```
````
I think it's worth adding one more block here with a couple of real examples, instead of just the `--help` usage.  Things like this:

```
> ghost-inspector test execute xxx  --apiKey xxx
```

justinklemm commented on bin/commands/folder/create.js:
```diff
@@ -0,0 +1,17 @@
+      const client = helpers.getClient(argv)
+      const result = await client.createFolder(argv.organizationId, argv.folderName)
+      helpers.print(result)
+    } catch (error) {
+      throw error
+    }
+    process.exit(0)
```
I haven't gone through and tracked down where these exceptions end up (so maybe you're doing this already), but if this request blows up then I think we should do `process.exit(1)`. We should even consider doing it if the [API returns an error](https://ghostinspector.com/docs/api/errors/) (which I know weirdly have 200 statuses, so we'd have to do some checking of the response).

The idea is that these are often going to be used in CI pipelines and if the command fails, it should blow up.

aaronfay replied:
Yeah, if we throw an error `yargs` will do the right thing already:

```
❯ ghost-inspector folder get 1234
ghost-inspector folder get <folder-id>

Fetch a single folder.

Options:
  --version  Show version number                                                                               [boolean]
  --apiKey   Your Ghost Inspector API key.              [required] [default: "xxxx"]
  --help     Show help                                                                                         [boolean]

Error: Some error
    at Object.handler (/Users/magius/workspace/node-ghost-inspector/bin/commands/folder/get.js:8:11)
    at Object.runCommand (/Users/magius/workspace/node-ghost-inspector/node_modules/yargs/lib/command.js:242:26)
    at Object.parseArgs [as _parseArgs] (/Users/magius/workspace/node-ghost-inspector/node_modules/yargs/yargs.js:1087:30)
    at Object.runCommand (/Users/magius/workspace/node-ghost-inspector/node_modules/yargs/lib/command.js:200:45)
    at Object.parseArgs [as _parseArgs] (/Users/magius/workspace/node-ghost-inspector/node_modules/yargs/yargs.js:1087:30)
    at Object.get [as argv] (/Users/magius/workspace/node-ghost-inspector/node_modules/yargs/yargs.js:1021:21)
    at Object.<anonymous> (/Users/magius/workspace/node-ghost-inspector/bin/cli.js:12:3)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)

node-ghost-inspector git/aaronfay/ch4307/add-cli-commands-to-node-library*
❯ echo $?
1
```

justinklemm reviewed (commented):
Code looks great 👌 

I added a few notes throughout. One thought:

It looks like the commands just execute the API call and return the JSON. But thinking ahead to how this will be used, it seems like that still leaves a lot for the user to do... For instance, if I execute a suite, how do I know what happened? I just get a big JSON dump. I've still got to parse it with `jq` or `grep` it or do some other ugliness with bash to output it and make sense of what happened... Ideally, I just want to add `npx ghost-inspector suite execute xxx ` to my CI config and be done.

It feels like we should have some kind of universal option that actually parses the JSON and displays some human readable output, then does `process.exit(0)` or `process.exit(1)` for passing or failing, respectively. Maybe that should even be the default. I'd prefer to see that in the command line myself, versus a JSON dump. The default output would look similar to what you get when you run `mocha` tests and then we add a `--json` option to revert back to what we're doing right now. Thoughts?

aaronfay commented:
Yeah, I originally had it like that, if a test fails then `exit(1)` but then changed it thinking, what if the user wants to pass/fail based on the screenshot, or both? It wouldn't take a whole lot to add a few options to tweak the status code based on the test pass/fail (which totally makes sense in CI), but if you don't return JSON with `execute` then do we run into this scenario where some commands return JSON and others don't? Like what would `ghost-inspector suite get` do?

Maybe it makes sense to add a flag `--exitOnFail` and `--exitOnFailScreenshot` or something for the execute commands?

aaronfay commented:
This should be good to go 👍🏻 

I've added unit tests for the CLI handlers, there's not a lot of logic in there, but enough to justify having something in place. I normally dislike abstracting test "helpers" but in this case all the command handlers are so consistent I think the complexity is worth making these a little easier to scan and read. There are extra tests for four handlers, `test execute`, `test execute-on-demand`, `suite execute`, and `result wait`. These handle scenarios where a user might use `--errorOnFail` or `--errorOnScreenshotFail` to adjust the exit status based on the test/suite result.

aaronfay commented:
I've just added the last of the tests to make sure the responses for execution with multiple results is working as expected. We should be good with this branch now.

justinklemm reviewed (approved):
👏 👏 👏  Awesome work. Code looks good to me and +1 for the extra tests.

I think we can merge this down at your convenience. I can then do a quick dependency update and push a new major version.

simpixelated commented on bin/commands/folder/get.js:
```diff
@@ -0,0 +1,24 @@
+        })
+      }
+    } catch (error) {
+      throw error
```
What's the point of the `try`/`catch` wrapper on these if it just throws? Placeholder for handling it different in the future?

aaronfay replied:
Ah good catch, I was unwrapping the error but changed my mind and never took out the try/catch. I'll make a note to pull them out at some point.

justinklemm replied:
I just went ahead and removed them in my dep updates PR here: https://github.com/ghost-inspector/node-ghost-inspector/pull/27

aaronfay replied:
🙏🏻
